packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -236,7 +365,101 @@ contract MetaTransactionWallet is
   }
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+   * @notice Executes a refundable meta-transaction on behalf of the signer.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param maxGasPrice The maximum gas price the user is willing to pay.
+   * @param gasLimit The gas limit for entire relayed transaction.
+   * @param metaGasLimit The gas limit for just the meta-transaction.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The return value of the meta-transaction execution.
+   */
+  function executeMetaTransactionWithRefund(
+    address destination,
+    uint256 value,
+    bytes calldata data,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit >= gasleft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
+    require(msg.sender == tx.origin, "relayer must be EOA");
+    if (data.length > 0) require(Address.isContract(destination), "Invalid contract address");
+    require(metaGasLimit < gasLimit, "metaGasLimit must be less than gasLimit");
+
+    // checking gasCurrency is out of scope, would require pre-compile
+    {
+      address _signer = getMetaTransactionWithRefundSigner(
+        destination,
+        value,
+        data,
+        nonce,
+        maxGasPrice,
+        gasLimit,
+        metaGasLimit,
+        v,
+        r,
+        s
+      );
+      require(_signer == signer, "Invalid meta-transaction signer");
+    }
+
+    nonce = nonce.add(1);
+
+    bytes memory returnData;
+
+    {
+      // TODO: ask Contracts about a more accurate estimate
+      uint256 gasAfterMTxExecution = 7900; // estimated by emitting gasleft() at desired benchmark times
+
+      if (
+        address(this).balance >=
```

Comment:
If I understand correctly, nothing could have happened since the balance check at the start of this function that would change the contract balance. If this is true, this balance check feels redundant.